modules/services/window-managers/river/default.nix

```diff
@@ -0,0 +1,225 @@
+      mapAttrsToList (option: value:
+        let
+          # fooBarOption -> foo-bar-option
+          formatOptionName =
+            builtins.replaceStrings upperChars (map (c: "-${c}") lowerChars);
+
+        in "riverctl input ${inputName} ${formatOptionName option} ${
+          toString value
+        }") (filterAttrs (n: v: v != null) inputConfig)) cfg.config.inputs;
+  in concatStringsSep "\n" (lists.flatten inputSettings);
+
+  inputs = ''
+    ${keyboardLayout}
+
+    ${inputsSettings}
+  '';
+
+  # CONFIGURATION
+  convertColor = colorString: "0x${replaceStrings [ "#" ] [ "" ] colorString}";
+  xCursorTheme = let
+    themeName = cfg.config.xCursorTheme.name;
+    size = cfg.config.xCursorTheme.size;
+  in optionalString (themeName != "") "riverctl xcursor-theme ${themeName}${
+    optionalString (size != null) " ${toString size}"
+  }";
+
+  configuration = ''
+    riverctl attach-mode ${cfg.config.attachMode}
+    riverctl background-color ${convertColor cfg.config.backgroundColor}
+    riverctl border-color-focused ${
+      convertColor cfg.config.border.color.focused
+    }
+    riverctl border-color-unfocused ${
+      convertColor cfg.config.border.color.unfocused
+    }
+    riverctl border-color-urgent ${convertColor cfg.config.border.color.urgent}
+    riverctl border-width ${toString cfg.config.border.width}
+    riverctl focus-follows-cursor ${cfg.config.focusFollowsCursor}
+    riverctl hide-cursor timeout ${toString cfg.config.hideCursor.timeout}
+    riverctl hide-cursor timeout ${
+      if cfg.config.hideCursor.whenTyping then "enabled" else "disabled"
+    }
+    riverctl set-cursor-warp ${cfg.config.cursorWarp}
+    riverctl set-repeat ${cfg.config.repeatRate}
+    ${xCursorTheme}
+  '';
+
+  formatStartupProgram = command: ''riverctl spawn "${command}"'';
+
+in {
+  meta.maintainers = [ maintainers.GaetanLepage ];
+
+  options = import ./options.nix { inherit pkgs lib; };
+
+  config = mkIf cfg.enable {
+    assertions = [
+      (hm.assertions.assertPlatform "wayland.windowManager.river" pkgs
+        platforms.linux)
+    ];
+
+    home.packages = optional (cfg.package != null) cfg.package
+      ++ optional cfg.xwayland pkgs.xwayland;
+
+    xdg.configFile."river/init".source = pkgs.writeShellScript "init" ''
+      ### This file was generated with Nix. Don't modify this file directly.
+
+      ### SHELL VARIABLES ###
+      ${config.lib.shell.exportAll cfg.extraSessionVariables}
+
+      ### RULES ###
+      ${rules}
+
+      ### MAPPINGS ###
+      ${mappings}
+
+      ### CONFIGURATION ###
+      ${configuration}
+
+      ### INPUTS ###
+      ${inputs}
+
+      ### EXTRA CONFIGURATION ###
+      ${cfg.extraConfig}
+
+      ### STARTUP PROGRAMS ###
+      ${concatMapStringsSep "\n" formatStartupProgram cfg.startupPrograms}
+
+      ### LAYOUT ###
+      riverctl default-layout ${cfg.config.layoutGenerator.name}
+      ${cfg.config.layoutGenerator.name} ${cfg.config.layoutGenerator.arguments} &
```

Comment:
systemd integration should follow:

```
### SYSTEMD ###
${pkgs.dbus}/bin/dbus-update-activation-environment --systemd DISPLAY WAYLAND_DISPLAY XDG_CURRENT_DESKTOP XDG_SESSION_TYPE XDG_VTNR NIXOS_OZONE_WL
systemctl --user --no-block start river-session.target
```

with a conditional for `cfg.systemdIntegration` of course.

I based this on the Sway systemd integration, <https://github.com/nix-community/home-manager/blob/0ee5ab611dc1fbb5180bd7d88d2aeb7841a4d179/modules/services/window-managers/i3-sway/sway.nix#L320>, and tested it on my machine.

Reply:
We should probably make that bit generic across Wayland window managers in a different PR

Reply:
I added the proper systemd `dbus` call.